### PR TITLE
Strengthen state tracking in const-prop

### PR DESCRIPTION
**Pull request by cjgillot**
Some/many of the changes are replicated between both the const-prop lint and the const-prop optimization.

Behaviour changes:
- const-prop opt does not give a span to propagated values. This was useless as that span's primary purpose is to diagnose evaluation failure in codegen.
- we remove the `OnlyPropagateInto` mode. It was only used for function arguments, which are better modeled by a write before entry.
- the tracking of assignments and discriminants make clearer that we do nothing in `NoPropagation` mode or on indirect places.

### CODE REVIEW

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
r=me with operand_from_scalar havings its span arg removed

File: compiler/rustc_mir_transform/src/const_prop.rs [hunk 1]
```diff
@@ -1027,23 +964,17 @@ impl<'tcx> MutVisitor<'tcx> for ConstPropagator<'_, 'tcx> {
...
-                assert!(
-                    self.get_const(local.into()).is_none()
-                        || self
-                            .layout_of(self.local_decls[local].ty)
-                            .map_or(true, |layout| layout.is_zst())
-                )
+        let can_const_prop = std::mem::take(&mut self.ecx.machine.can_const_prop);
+        for (local, &mode) in can_const_prop.iter_enumerated() {
+            match mode {
+                ConstPropMode::FullConstProp => {}
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/const_prop.rs:970]: It looks to me like the regression is from this loop? At least that's what my profiler indicates. It looks like basically all the cycles in `ConstProp` are driving this iterator and hitting this first case:
```
 16.79 │ d0:┌─→inc    %rdi
       │    │<core::slice::iter::Iter<rustc_mir_transform::const_prop::ConstPropMode> as core::iter::traits::iterator::Iterator>::next:
 16.05 │    │  add    $0x48,%rdx
 16.52 │    │  cmp    %rdi,%rax
       │    │↓ je     12e
       │    │<rustc_middle::mir::Local>::from_usize: 
       │    │} 
       │    │
       │    │///////////////////////////////////////////////////////////////////////////
       │    │// Variables and temps 
       │    │ 
       │    │rustc_index::newtype_index! {
 25.17 │ dc:│  cmp    %r8,%rdi
       │    │↓ je     1b7
       │    │<rustc_mir_transform::const_prop::ConstPropagator as rustc_middle::mir::visit::MutVisitor>::visit_basic_block_data: 
       │    │match mode {
 25.18 │    ├──cmpb   $0x1,(%rcx,%rdi,1)
       │    └──jne    d0
```

Comment (saethlin) [reply compiler/rustc_mir_transform/src/const_prop.rs:970]: I will not be working on this, but I suspect the problem centers around there being a _lot_ of `FullConstProp` in here and we could be iterating over less elements somehow.

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/const_prop.rs:970]: Thanks for the analysis. I'll handle it after hours.

**Pull request merged**